Add test zones for Address03 (#1354)
dotse/zonemaster: docs/public/specifications/test-zones/Address-TP/address03.md

> # Specification of test zones for ADDRESS03

Author: Good idea. I’ll use DNSSEC10 as a template.

| INCOMPLETE-PTR-1      | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| INCOMPLETE-PTR-2      | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| NON-MATCHING-NAMES    | NAMESERVER_IP_PTR_MISMATCH    | 2)                      |
| PTR-IS-GOOD-CNAME-1   | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| PTR-IS-GOOD-CNAME-2   | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| PTR-IS-DANGLING-CNAME | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| PTR-IS-ILLEGAL-CNAME  | NAMESERVER_IP_WITHOUT_REVERSE | NAMESERVER_IP_PTR_MATCH |

* (2) All tags except for those specified as “Mandatory message tags”

## Zone setup for test scenarios

Assumptions for the scenario specifications:
* For each scenario zone there are two name servers configured.
* Unless stated otherwise, all name servers respond as follows:
  * Responds with a SOA record for the zone on query for SOA.
  * All responses are authoritative with [RCODE Name] "NoError".
  * EDNS, version 0, is included in all responses on queries with EDNS.
  * EDNS is not included in responses on queries without EDNS.

>   * Unknown EDNS option codes are not included in responses.

Author: Done.

### ALL-NS-HAVE-PTR-1
* Zone: "all-ns-have-ptr-1.address03.xa."

>   * A happy path: a zone is delegated to two name servers with IPv4 and IPv6 addresses which have single and correct PTR records.

Author: Sounds good. I’ll implement that.

* NAMESERVER_IP_PTR_MISMATCH
* NAMESERVER_IP_WITHOUT_REVERSE
* NO_RESPONSE_PTR_QUERY

## Test scenarios and message tags

If a message tag is not listed for the scenario, its presence or non-presence is
irrelevant to the test scenario and must be ignored.

| Scenario name         | Mandatory message tags        | Forbidden message tags  |
|:----------------------|:------------------------------|:------------------------|
| ALL-NS-HAVE-PTR-1     | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| ALL-NS-HAVE-PTR-2     | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| NO-NS-HAVE-PTR        | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| INCOMPLETE-PTR-1      | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| INCOMPLETE-PTR-2      | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |
| NON-MATCHING-NAMES    | NAMESERVER_IP_PTR_MISMATCH    | 2)                      |
| PTR-IS-GOOD-CNAME-1   | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| PTR-IS-GOOD-CNAME-2   | NAMESERVER_IP_PTR_MATCH       | 2)                      |
| PTR-IS-DANGLING-CNAME | NAMESERVER_IP_WITHOUT_REVERSE | 2)                      |

> | PTR-IS-ILLEGAL-CNAME  | NAMESERVER_IP_WITHOUT_REVERSE | NAMESERVER_IP_PTR_MATCH |

Author: Good point, actually: I’m missing a scenario where a name server has a PTR record in an inverse zone whose name servers that either SERVFAIL or fail to respond. I’ll add that.

> | PTR-IS-ILLEGAL-CNAME  | NAMESERVER_IP_WITHOUT_REVERSE | NAMESERVER_IP_PTR_MATCH |

Reviewer: My comment in test-zone-output.md is related to this comment.

> | PTR-IS-ILLEGAL-CNAME  | NAMESERVER_IP_WITHOUT_REVERSE | NAMESERVER_IP_PTR_MATCH |

Author: Yes, I’ve moved the statement about `NO_RESPONSE_PTR_QUERY` from `test-zones-output.md` to the test scenario specification.

* (1) All tags except for those specified as "Forbidden message tags" (no instances for these test scenarios)
* (2) All tags except for those specified as "Mandatory message tags"

## Test scenarios and setup of test zones

### Default zone configuration
Unless otherwise specified in the specific scenario specification, the test zone
for the scenario will follow the default setup as stated below. The `child zone`
is the zone to be tested for the scenario.

* The child zone is `SCENARIO.address03.xa`.

>   * There is a zone file for the child zone.

Author: Done.

>   * The authoritative name servers for the zone all have an IPv4 and IPv6

Author: Done.

    address, and the reverse zones contain a single PTR resource record
    matching their names for all of their addresses.
  * The NS record set in the child zone is consistent with the parent zone’s
    delegation.
  * The authoritative name servers all serve the same authoritative data.
* The parent zone is `address03.xa`.
  * It is served by two in-bailiwick NS (ns1.address03.xa and
    ns2.address03.xa).
  * ns1 and ns2 have the same zone content.

>   * ns1 and ns2 have both IPv4 and IPv6 glue.

Author: Done.

>   * The records matching glue in the zone are complete.

Author: Done.

* All authoritative name servers for the scenario’s child zones have names
  matching ns\<NUMBER\>.child.address03.xa. These name servers’s names are
  abbreviated by leaving out address03.xa from their names.
* All responses will have the AA bit set.
* All responses will have the [RCODE Name] "NoError".

### ALL-NS-HAVE-PTR-1

> A happy path: a zone is delegated to two name servers with IPv4 and IPv6 addresses which have single and correct PTR records.

Author: Sure.

* Zone: all-ns-have-ptr-1.address03.xa
  * Delegated to: ns1.child and ns2.child.

### ALL-NS-HAVE-PTR-2

Another happy path: like ALL-NS-HAVE-PTR-1, but one of the name server’s IP addresses (both IPv4 and IPv6) has multiple PTR records, among which one is consistent with the name server’s name.

* Zone: all-ns-have-ptr-2.address03.xa
  * Delegated to: ns1.child and ns3.child.
  * ns3.child’s IP addresses have multiple PTR records, among which one points
    to ns3.child.

### NO-NS-HAVE-PTR

> None of the name server’s IP addresses have PTR records at all. One of them has a PTR query return NODATA, the other NXDOMAIN.

Author: I’ll partially apply this suggestion: I’m pretty sure English uses plural forms for zero, not singular (as would happen in French).

* Zone: no-ns-have-ptr.address03.xa
  * Delegated to: ns4.child and ns5.child.
  * ns4.child’s IP addresses have no PTR records; the reverse zone is
    configured to provoke NODATA responses on PTR queries by making the
    expected node an empty non-terminal.
  * ns5.child’s IP addresses have no PTR records; the reverse zone is
    configured to provoke NXDOMAIN responses on PTR queries.

### INCOMPLETE-PTR-1

> One of the name servers has PTR records for its IPv6 address, but not for its IPv4 address.

Author: Done.

* Zone: incomplete-ptr-1.address03.xa
  * Delegated to: ns1.child and ns6.child.
  * ns6.child’s IPv4 address has no PTR record, but its IPv6 address does.

### INCOMPLETE-PTR-2

> One of the name servers has PTR records for its IPv4 address, but not for its IPv6 address.

Author: Done.

* Zone: incomplete-ptr-2.address03.xa
  * Delegated to: ns1.child and ns7.child.
  * ns7.child’s IPv4 address has a PTR record, but its IPv6 address does not.

### NON-MATCHING-NAMES

> One of the name servers has PTR records pointing to a different hostname.

Author: Actually this one is slightly wrong because both name server’s IP addresses are configured that way. One has more than one PTR, the other has only one PTR. I’ve fixed that.

* Zone: non-matching-names.address03.xa
  * Delegated to: ns8.child and ns9.child.

>   * ns8.child’s IP addresses have a single PTR record, but its hostname is

Reviewer: ```suggestion
  * ns8.child’s IP addresses have a single PTR record, but its hostname in RDATA is

```

Author: Yes, good point.

    different from the name server’s name.
  * ns9.child’s IP addresses have more than one PTR record, and each of them

>     has a hostname different from the name server’s name.

Reviewer: ```suggestion
    has a hostname in RDATA different from the name server’s name.

```

Author: Sure.

### PTR-IS-GOOD-CNAME-1

One of the name servers has PTR records pointing to an alias whose target is in the same reverse zone.

* Zone: ptr-is-good-cname-1.address03.xa
  * Delegated to: ns1.child and ns10.child.

>   * ns10.child’s IP addresses have reverse entries that are aliased to another
>     node in the same zone. In other words, resolving the PTR resource records

Reviewer: ```suggestion
  * ns10.child’s IP addresses have reverse names that are aliased (CNAME) to another
    name in the same zone. In other words, resolving the PTR resource records
```

Author: Done.

    for their IP addresses returns a CNAME resource record and the PTR record
    after walking the CNAME chain.

### PTR-IS-GOOD-CNAME-2

> One of the name servers has PTR records pointing to an alias whose target lives in a different zone.

Author: Done.

* Zone: ptr-is-good-cname-2.address03.xa
  * Delegated to: ns1.child and ns11.child.

>   * ns11.child’s IP addresses have reverse entries that are aliased to another
>     node in a different zone. In other words, resolving the PTR resource
>     records for their IP addresses returns only a CNAME resource record,
>     therefore Zonemaster needs to do another query for the name at the target
>     of the CNAME resource record.

Reviewer: ```suggestion
  * ns11.child’s IP addresses have reverse names that are aliased (CNAME) to another
    name in a different zone. In other words, resolving the PTR resource
    records for their IP addresses returns only a CNAME resource record,
    and another query for the name at the target of the CNAME resource record
    is needed.
```

Author: Better indeed.

### PTR-IS-DANGLING-CNAME

> One of the name servers has PTR records pointing to an alias whose target does not exist.

Author: Done.

* Zone: ptr-is-dangling-cname.address03.xa
  * Delegated to: ns5.child and ns12.child.
  * ns5.child is configured as described in the NO-NS-HAVE-PTR scenario.

>   * ns12.child’s IP addresses have reverse entries that are aliased to a
>     nonexistent node. In other words, there is a CNAME pointing to a node
>     that does not exist.

Reviewer: ```suggestion
  * ns12.child’s IP addresses have reverse names that are aliased (CNAME) to a
    nonexistent name. In other words, there is a CNAME pointing to a name
    that does not exist.
```

Author: Done.

### PTR-IS-ILLEGAL-CNAME

> One of the name servers has PTR records pointing to an alias with more than just one CNAME resource record.

Reviewer: ```suggestion
The reverse name of one of the name servers' IP address has multiple CNAME
records.
```

Author: Yes, very poor wording on my part. But the suggestion is incorrect: both the IPv4 and IPv6 addresses of the affected name servers have reverse names containing 2 CNAME RRs.

* Zone: ptr-is-illegal-cname.address03.xa
  * Delegated to: ns4.child and ns13.child.
  * ns4.child is configured as described in the NO-NS-HAVE-PTR scenario.

>   * ns13.child’s IP addresses have reverse entries that give two CNAME
>     resource records when queried for the PTR type, which is an illegal
>     configuration of the zone.

Reviewer: ```suggestion
  * ns13.child’s IP addresses have reverse names that give two CNAME
    resource records.
```
